src/main/java/org/zeplinko/commons/lang/ext/core/Retryable.java

```
@@ -0,0 +1,225 @@
+    public Retryable<R> retries(int maxRetries) {
+        this.maxRetries = Math.min(Math.max(0, maxRetries), MAX_RETRIES_UPPER_BOUND);
+        return this;
+    }
+
+    /**
+     * Sets the base delay that is applied before scheduling the next attempt.
+     * <p>
+     * The supplied {@code unit} must be {@link TimeUnit#SECONDS} or a unit with a
+     * finer granularity (e.g. {@code MILLISECONDS}). The effective duration must
+     * not exceed 60 seconds in total. Violating either constraint results in an
+     * {@link IllegalArgumentException}.
+     * </p>
+     *
+     * @param duration amount of time to wait between attempts
+     * @param unit     unit for the duration; seconds or finer
+     * @return this instance for method chaining
+     * @throws IllegalArgumentException if the constraints are violated
+     */
+    public Retryable<R> delay(long duration, @Nonnull TimeUnit unit) {
+        Objects.requireNonNull(unit);
+        if (unit.compareTo(TimeUnit.SECONDS) > 0) {
+            throw new IllegalArgumentException("TimeUnit must be SECONDS or a more precise unit");
+        }
+        if (unit.toSeconds(duration) > 60) {
+            throw new IllegalArgumentException("Delay cannot exceed 60 seconds");
+        }
+        this.baseDelayMillis = unit.toMillis(duration);
+        return this;
+    }
+
+    /**
+     * Selects the back-off strategy to compute the delay before the next retry.
+     *
+     * @param strategy the {@link Backoff} strategy to apply
+     * @return this instance for method chaining
+     */
+    public Retryable<R> backoff(@Nonnull Backoff strategy) {
+        this.backoffStrategy = Objects.requireNonNull(strategy);
+        return this;
+    }
+
+    /**
+     * Executes the configured task applying all retry rules and validation.
+     *
+     * @return a {@link Result} containing either the successful value or the
+     *         exception ultimately thrown/raised
+     */
+    public Result<R, Exception> run() {
```

We need to restrict that run can be invoked only once. Post that it should throw an exception.

---

`Result<R, Exception>` is technically `Try<R>`

But I don't think the result here can be represented with just 2 signals.
We might need 3 signals here
1. SuccessValue
2. Failure due to exception and retryOnFailure returning false
3. Failure due to attempts being exhausted 

Thoughts?

---

> We need to restrict that run can be invoked only once. Post that it should throw an exception.
Originally i thought of modeling the Retryable after `AbstractOutcome` with a RetryableOutcome as the result (of run()) and a `RetryableBuilder` to collect the parameter.

This seems a bit unnecessary and I wanted to keep things simple here. The Retryable is actually a builder class which once is completely build offloads the result to the Result class. (`run` is similar to `build`)

Once the builder builds (`run()`), there is not more builder and hence it cannot be executed twice.

---

hmmm... the Retryable instance can be kept separately and people can invoke run multiple times making it multi-use.

So lets say someone did this
```java
Retryable rt = Retryable.of( ()-> false )
.retryUntilResut(r -> true )
.maxRetries(100);

Result r1 = rt.run();
Resutl r2 = rt.run();
```

Here we cannot maintain any state in Retryable. So what if instead of Result we return a `RetryableOutcome` as the result of run, with all the states. It has the result, attempts, why if failed (attempts exhausted, exceptions etc).

That way the Retryable can implement a callable as well.

I didn't want to keep any states inside Retrable since its being modeled as a builder.

---

thoughts? @ShivamNagpal

---

@a-anand-91119 
Considering all the comments from you, I guess this is what we can do here

1. Keep Retryable as a stateless class as it is now.
2. Change the run method to return RetryableOutcome<R> class. This class can be static nested class inside the Retryable.
3. RetryableOutcome will represent the following. (This is Java 8 still, so no sealed class is available here, we will have to use regular class only)
   1. Whether outcome is success or not
   2. Number of attempts
   3. If outcome is failure then what is the reason for it.
      1. Attempt Exhaustion (can be due to success predicate non satisfaction or due to failure predicate satisfaction for all the runs)
      2. Abort due to failure predicate dissatisfaction.

What do you think?

---

Yeah. Lets do that. 

Note:  
- Calling the result class `Outcome` since we use it any way like `Retyable.Outcome`. `Retryable.RetryableOutcome` would be redundant.
- Also implemented Callable, that way this can be used with executors and futures.